libs/go/aealite/protocols/dialogue.go

```diff
@@ -0,0 +1,437 @@
+	isMessageBySelf(AbstractMessage) bool
+	isMessageByOther(AbstractMessage) bool
+	getMessage(MessageId) AbstractMessage
+	hasMessageId(MessageId) bool
+
+	update(AbstractMessage)
+	validateNextMessage(AbstractMessage) (bool, string)
+	basicValidations(AbstractMessage) (bool, string)
+	basicValidationInitialMessage(AbstractMessage) (bool, string)
+	basicValidationNonInitialMessage(AbstractMessage) (bool, string)
+	isEmpty() bool
+	updateIncomingAndOutgoingMessages(AbstractMessage)
+	isBelongingToADialogue(AbstractMessage) bool
+
+	validateMessageTarget(AbstractMessage) string
+	validateMessageId(AbstractMessage) string
+	getMessageById(MessageId) AbstractMessage
+	getOutgoingNextMessageId() int
+	getIncomingNextMessageId() int
+	updateDialogueLabel(DialogueLabel)
+	customValidation(AbstractMessage) (bool, string)
+	getStringRepresentation() string
+}
+type AbstractMessageInterface interface {
+	validPerformatives() []string
+	hasSender() (Address, error)
+	hasCounterparty() (Address, error)
+	dialogueReference() [2]string
+	messageId() MessageId
+	performative() Performative
+	target() Target
+}
+
+func (message AbstractMessage) hasSender() (Address, error) {
+	if message.sender != "" {
+		return message.sender, nil
+	}
+	return "", errors.New("Sender address does not exist.")
+}
+
+func (message AbstractMessage) hasCounterparty() (Address, error) {
+	if message.to != "" {
+		return message.to, nil
+	}
+	return "", errors.New("Counterparty address does not exist.")
+}
+
+type DialogueLabel struct {
+	dialogueReference       [2]string
+	dialogueOpponentAddress Address
+	dialogueStarterAddress  Address
+}
+
+type Dialogue struct {
+	dialogueLabel     DialogueLabel
+	dialogueMessage   AbstractMessage
+	selfAddress       Address
+	outgoingMessages  []AbstractMessage
+	incomingMessages  []AbstractMessage
+	lastMessageId     int
+	orderedMessageIds []int
+	rules             RuleType
+}
+type AbstractMessage struct {
+	dialogueReference [2]string
+	messageId         MessageId
+	target            Target
+	performative      []string
+	message           []byte
+	to                Address
+	sender            Address
+}
+
+// store dialogues by opponent address
+var dialogueStorage = make(map[Address][]Dialogue)
+
+// store dialogue by dialogue label
+var dialogueByDialogueLabel = make(map[DialogueLabel]Dialogue)
+
+func (dialogue Dialogue) update(message AbstractMessage) {
+	if message.sender == "" {
+		message.sender = dialogue.selfAddress
+	}
+	messageExistence := dialogue.isBelongingToADialogue(message)
+	if !messageExistence {
+		fmt.Println("Error: message does not exist to this dialogue")
+	}
+
+	// TODO
+	// validate next message
+	// check if dialogue message is valid
+	dialogue.validateNextMessage(message)
+	dialogue.updateIncomingAndOutgoingMessages(message)
+}
+
+func (dialogue Dialogue) validateNextMessage(message AbstractMessage) (bool, string) {
+	is_basic_validated, msg_basic_validation := dialogue.basicValidation(message)
```

Comment:
in Golang, identifiers should be camelCase:
```suggestion
	isBasicValidated, msgBasicValidation := dialogue.basicValidation(message)
```